File: src/solutions/2023/05/csharp/Main.cs

using Core;

namespace Year2023
{
   public class Day05
   {
      static SolutionOptions OPTIONS = new(
      )
      {
         HasAlternate = false,
         HasIO = false,
      };

      private static (List<long[]>, IEnumerable<IEnumerable<((long, long), (long, long))>>) ParseInput(string input, bool single)
      {
         string[] parsed = input.Split("\n\n");
         long[] header = parsed[0]
            .Split(':')[1]
            .Split(' ')
            .Where((str) => str != "")
            .Select(long.Parse).ToArray();
         List<long[]> seedRanges;
         if (single)
            seedRanges = header.Select(v => new long[] { v, v }).ToList();
         else
            seedRanges = new long[header.Length / 2]
               .Select((_, idx) => new long[] { header[idx * 2], header[idx * 2] + header[idx * 2 + 1] })
               .ToList();

         IEnumerable<IEnumerable<((long, long), (long, long))>> srcToDestRanges = parsed[1..].Select(
            (p) =>
               p
                  .Split('\n')[1..]
                  .Select((str) => str.Split(' ').Select(long.Parse).ToArray())
                  .Select((v) => ((v[1], v[1] + v[2] - 1), (v[0], v[0] + v[2] - 1)))
         );

         return (seedRanges, srcToDestRanges);
      }

      static string Solve(string input, bool single)
      {
         var (seedRanges, srcToDestRanges) = ParseInput(input, single);

         foreach (var groups in srcToDestRanges)
         {
            foreach (var g in groups)
            {
               List<long[]> newSeeds = new List<long[]>();
               foreach (var r in seedRanges)
               {
                  if (r[0] < g.Item1.Item1 && g.Item1.Item1 < r[1])
                  {
                     newSeeds.Add([r[0], g.Item1.Item1 - 1]);
                     r[0] = g.Item1.Item1;
                  }
                  if (r[0] < g.Item1.Item2 && g.Item1.Item2 < r[1])
                  {
                     newSeeds.Add([g.Item1.Item2 + 1, r[1]]);
                     r[1] = g.Item1.Item2;
                  }
               }
               seedRanges.AddRange(newSeeds);
            }
            foreach (var r in seedRanges)
            {
               ((long, long), (long, long))? found = null;
               try
               {
                  found = groups.First(
                  (g) =>
                     g.Item1.Item1 <= r[0] &&
                     r[0] <= g.Item1.Item2 &&
                     r[1] >= g.Item1.Item1 &&
                     g.Item1.Item2 >= r[1]
               );
               }
               catch { }
               if (found != null)
               {
                  var diff = found.Value.Item2.Item1 - found.Value.Item1.Item1;
                  r[0] += diff;
                  r[1] += diff;
               }
            }
         }

         return seedRanges.Select(x => x[0]).Min().ToString();
      }

      static string Part1(string input, bool isTest = false)
      {
         return Solve(input, true);
      }

      static string Part2(string input, bool isTest = false)
      {
         return Solve(input, false);
      }

      static void Main(string[] args)
      {
         Runner.Run(args, Part1, Part2, OPTIONS);
      }
   }
}

File: src/solutions/2023/05/rust/main.rs
extern crate aoc_core;
use std::{env, iter::FromIterator};

fn get_options() -> aoc_core::options::SolutionOptions {
   aoc_core::options::SolutionOptions {
      ..Default::default()
   }
}

fn parse_input(input: &str, single: bool) -> (Vec<(i64, i64)>, Vec<Vec<((i64, i64), (i64, i64))>>) {
   let mut parsed = input.split("\n\n");
   let header = parsed
      .next()
      .unwrap()
      .split(':')
      .next_back()
      .unwrap()
      .split(' ')
      .filter(|x| *x != "")
      .map(|x| x.parse::<i64>().unwrap())
      .collect::<Vec<i64>>();

   let seed_ranges: Vec<(i64, i64)> = if single {
      header
         .iter()
         .map(|x| (x.clone(), x.clone()))
         .collect::<Vec<(i64, i64)>>()
   } else {
      header
         .iter()
         .enumerate()
         .filter(|(idx, _)| idx % 2 == 0)
         .map(|(idx, x)| (*x, *x + header[idx + 1]))
         .collect::<Vec<(i64, i64)>>()
   };

   let src_to_dest_ranges = parsed
      .into_iter()
      .map(|x| {
         x.lines()
            .skip(1)
            .map(|str| {
               str.split(' ')
                  .map(|x| x.parse::<i64>().unwrap())
                  .collect::<Vec<i64>>()
            })
            .map(|v| ((v[1], v[1] + v[2] - 1), (v[0], v[0] + v[2] - 1)))
            .collect::<Vec<((i64, i64), (i64, i64))>>()
      })
      .collect();

   (seed_ranges, src_to_dest_ranges)
}

fn solve(input: &str, single: bool) -> String {
   let (seed_ranges, src_to_dest_ranges) = parse_input(input, single);
   let seed_ranges = &mut Vec::from_iter(seed_ranges);
   for groups in src_to_dest_ranges.iter() {
      for g in groups.iter() {
         let mut new_seeds = Vec::new();
         for r in seed_ranges.into_iter() {
            if r.0 < g.0 .0 && g.0 .0 < r.1 {
               new_seeds.push((r.0, g.0 .0 - 1));
               r.0 = g.0 .0;
            }
            if r.0 < g.0 .1 && g.0 .1 < r.1 {
               new_seeds.push((g.0 .1 + 1, r.1));
               r.1 = g.0 .1;
            }
         }
         seed_ranges.append(&mut new_seeds);
      }
      for r in seed_ranges.into_iter() {
         let found = groups
            .iter()
            .find(|g| g.0 .0 <= r.0 && r.0 <= g.0 .1 && r.1 >= g.0 .0 && g.0 .1 >= r.1);
         if !found.is_none() {
            let f = found.unwrap();
            let diff = f.1 .0 - f.0 .0;
            r.0 += diff;
            r.1 += diff;
         }
      }
   }

   seed_ranges.iter().map(|x| x.0).min().unwrap().to_string()
}

fn part_1(input: &str, _is_test: bool) -> String {
   return solve(input, true);
}

fn part_2(input: &str, _is_test: bool) -> String {
   return solve(input, false);
}

fn main() {
   let args = Vec::from_iter(env::args());
   aoc_core::run(
      args.iter().map(|x| x.as_ref()).collect::<Vec<_>>(),
      part_1,
      part_2,
      get_options(),
   );
}
